[REVIEW] Adds in support for chunked writers to java [skip-ci]

revans2 opened the pull request:
Adds in support so when writing to Parquet or Orc we can do it one table at a time, instead of all at once.  It also lets us write the data into a memory buffer instead of a local file.

jlowe commented on java/src/main/native/src/TableJni.cpp:
```diff
@@ -22,12 +22,185 @@
+      long amount_to_copy = left_to_copy < buffer_amount_available ? left_to_copy : buffer_amount_available;
+      char * copy_to = current_buffer_data + current_buffer_written;
+
+      CUDA_TRY(cudaMemcpyAsync(copy_to, copy_from, amount_to_copy, cudaMemcpyDeviceToHost, stream));
```
Does the data sink API semantics guarantee it is safe to return before the data is copied?  In other words, we know that the entity calling the sink will synchronize on the stream passed to us after we return rather than implicitly relying on us to synchronize?

jlowe replied:
This comment was not addressed.  cc: @nvdbaranec to see if this behavior is expected and documented in the API.

jlowe commented on java/src/main/native/src/TableJni.cpp:
```diff
@@ -22,12 +22,185 @@
+    if (current_buffer != nullptr) {
+      if (is_stream_set) {
+        // only sync with the stream when we are going to use the data.
+        CUDA_TRY(cudaStreamSynchronize(stream));
```
Technically we don't want to synchronize with the entire stream backlog, just the point after we issued the copy.  Really should be synchronizing on an event here that was added to the stream immediately following the copy.  Something to consider in the future.

jlowe commented on java/src/main/native/src/TableJni.cpp:
```diff
@@ -22,12 +22,185 @@
+};
+
+template <typename STATE>
+class jni_x_writer_handle final {
```
Curious about the `x` in the name.  Would `jni_table_writer_handle` be more readable?

jlowe commented on java/src/main/native/src/TableJni.cpp:
```diff
@@ -327,9 +550,47 @@ JNIEXPORT void JNICALL Java_ai_rapids_cudf_Table_writeParquet(JNIEnv* env, jclas
+      reinterpret_cast<cudf::jni::native_parquet_writer_handle *>(j_state);
+
+  if (state->sink.get() != nullptr) {
+    long alloc_size = 1024 * 1024 * 10; // 10 MB minimum
```
This is the second time this constant appears, probably worth creating something like `constexpr long DEFAULT_WRITER_BUFFER_SIZE = 10 * 1024 * 1024;`

jlowe commented on java/src/main/native/src/TableJni.cpp:
```diff
@@ -327,9 +550,47 @@ JNIEXPORT void JNICALL Java_ai_rapids_cudf_Table_writeParquet(JNIEnv* env, jclas
+  if (state->sink.get() != nullptr) {
+    long alloc_size = 1024 * 1024 * 10; // 10 MB minimum
+    if (alloc_size < mem_size/2) {
+      alloc_size = mem_size/2;
```
could simplify this with `long alloc_size = std::max(DEFAULT_WRITER_BUFFER_SIZE, mem_size / 2)`

jlowe commented on java/src/main/native/src/TableJni.cpp:
```diff
@@ -327,9 +550,47 @@ JNIEXPORT void JNICALL Java_ai_rapids_cudf_Table_writeParquet(JNIEnv* env, jclas
+      reinterpret_cast<cudf::jni::native_parquet_writer_handle *>(j_state);
+  try {
+    write_parquet_chunked_end(state->state);
+    // TODO should we do something to be 100% sure this is deleted???
```
Asserting the `use_count` might not be a bad idea if we are positive no other pointers should be sharing the state at this point.

revans2 replied:
I was thinking more along the idea of exception handling and wanting to be 100% sure that we delete even if `write_parquet_chunked_end` throws.

jlowe commented on java/src/main/native/src/TableJni.cpp:
```diff
@@ -379,48 +640,122 @@ JNIEXPORT jlongArray JNICALL Java_ai_rapids_cudf_Table_readORC(
+  cudf::jni::native_orc_writer_handle * state =
+      reinterpret_cast<cudf::jni::native_orc_writer_handle *>(j_state);
+
+  if (state->sink.get() != nullptr) {
```
```suggestion
  if (state->sink) {
```

jlowe commented on java/src/main/native/src/TableJni.cpp:
```diff
@@ -379,48 +640,122 @@ JNIEXPORT jlongArray JNICALL Java_ai_rapids_cudf_Table_readORC(
+  if (state->sink.get() != nullptr) {
+    long alloc_size = 1024 * 1024 * 10; // 10 MB minimum
+    if (alloc_size < mem_size/2) {
+      alloc_size = mem_size/2;
```
`std::max` again

jlowe commented on java/src/main/native/src/TableJni.cpp:
```diff
@@ -327,9 +550,47 @@ JNIEXPORT void JNICALL Java_ai_rapids_cudf_Table_writeParquet(JNIEnv* env, jclas
+  cudf::jni::native_parquet_writer_handle * state =
+      reinterpret_cast<cudf::jni::native_parquet_writer_handle *>(j_state);
+
+  if (state->sink.get() != nullptr) {
```
```suggestion
  if (state->sink.get()) {
```

jlowe commented on java/src/test/java/ai/rapids/cudf/TableTest.java:
```diff
@@ -2126,16 +2123,63 @@ void testParquetWriteToFileNoNames() throws IOException {
+      }
+    }
+
+    @Override
```
Nit: This override is unnecessary since interface provides a default.

revans2 commented:
@jlowe I think I have addressed your review comments.

jlowe commented on java/src/main/native/src/CudfJni.cpp:
```diff
@@ -105,6 +105,69 @@ native_jobjectArray<jobject> contiguous_table_array(JNIEnv* env, jsize length) {
   return native_jobjectArray<jobject>(env, env->NewObjectArray(length, contiguous_table_jclass, nullptr));
 }
 
+static jclass host_memory_buffer_jclass;
```
```suggestion
static jclass Host_memory_buffer_jclass;
```

revans2 commented:
@jlowe I addressed the one item I missed, and I changed how we do `cudaStreamSynchronize`. It is simpler and I didn't see any performance impact in practice.

revans2 commented:
@kkraus14 could you merge this in now that it is approved?
